File: src/fmtstrformatter.cpp
#include "fmtstrformatter.h"

namespace Newsboat {

FmtStrFormatter::FmtStrFormatter()
	: rs_object(fmtstrformatter::bridged::create())
{
}

void FmtStrFormatter::register_fmt(char f, const std::string& value)
{
	fmtstrformatter::bridged::register_fmt(*rs_object, f, value);
}

std::string FmtStrFormatter::do_format(const std::string& fmt,
	unsigned int width)
{
	auto formatted = fmtstrformatter::bridged::do_format(*rs_object, fmt, width);
	return std::string(formatted);
}

} // namespace Newsboat

File: rust/libnewsboat-ffi/src/fmtstrformatter.rs
use libNewsboat::fmtstrformatter;

// cxx doesn't allow to share types from other crates, so we have to wrap it
// cf. https://github.com/dtolnay/cxx/issues/496
struct FmtStrFormatter(fmtstrformatter::FmtStrFormatter);

#[cxx::bridge(namespace = "Newsboat::fmtstrformatter::bridged")]
mod bridged {
    extern "Rust" {
        type FmtStrFormatter;

        fn create() -> Box<FmtStrFormatter>;

        fn register_fmt(fmt: &mut FmtStrFormatter, key: u8, value: &str);
        fn do_format(fmt: &mut FmtStrFormatter, format: &str, width: u32) -> String;
    }
}

fn create() -> Box<FmtStrFormatter> {
    Box::new(FmtStrFormatter(fmtstrformatter::FmtStrFormatter::new()))
}

fn register_fmt(fmt: &mut FmtStrFormatter, key: u8, value: &str) {
    fmt.0.register_fmt(key as char, value.to_string());
}

fn do_format(fmt: &mut FmtStrFormatter, format: &str, width: u32) -> String {
    fmt.0.do_format(format, width)
}
